src/main/java/io/aiven/kafka/auth/audit/Auditor.java

```diff
@@ -61,64 +57,49 @@ protected Auditor(final Logger logger) {
 
     @Override
     public void configure(final Map<String, ?> configs) {
-        final AuditorConfig auditorConfig = new AuditorConfig(configs);
+        auditorConfig = new AuditorConfig(configs);
         auditScheduler.scheduleAtFixedRate(
             this::dump,
             auditorConfig.getAggregationPeriodInSeconds(),
             auditorConfig.getAggregationPeriodInSeconds(),
             TimeUnit.SECONDS
         );
-
-        formatter = createFormatter(auditorConfig);
-    }
-
-    private AuditorDumpFormatter createFormatter(final AuditorConfig auditorConfig) {
-        final String grouping = auditorConfig.getAggregationGrouping();
-        if (PRINCIPAL_AND_SOURCE_IP.getConfigValue().equals(grouping)) {
-            return new PrincipalAndIpFormatter();
-        } else if (PRINCIPAL.getConfigValue().equals(grouping)) {
-            return new PrincipalFormatter();
-        } else {
-            throw new RuntimeException("Not implemented");
-        }
     }
 
     @Override
-    public void addActivity(final Session session,
-                            final Operation operation,
-                            final Resource resource,
-                            final boolean hasAccess) {
-
-        final AuditKey auditKey = AuditKey.fromSession(session);
+    public final void addActivity(final Session session,
+                                  final Operation operation,
+                                  final Resource resource,
+                                  final boolean hasAccess) {
         auditLock.lock();
         try {
-            auditStorage.compute(auditKey, (key, userActivity) ->
-                onUserActivity(Objects.isNull(userActivity)
-                        ? new UserActivity()
-                        : userActivity, operation, resource, hasAccess)
-            );
+            addActivity0(session, operation, resource, hasAccess);
         } finally {
             auditLock.unlock();
         }
     }
 
+    protected abstract void addActivity0(final Session session,
+                                         final Operation operation,
+                                         final Resource resource,
+                                         final boolean hasAccess);
+
     @Override
     public void stop() {
         auditScheduler.shutdownNow();
         try {
-            auditScheduler.awaitTermination(5, TimeUnit.SECONDS);
             dump();
+            auditScheduler.awaitTermination(5, TimeUnit.SECONDS);
         } catch (final InterruptedException e) {
-            // Intentionally ignored
+            logger.warn("Couldn't dump latest messages", e);
```

Comment:
This log message is incorrect: `InterruptedException` might be thrown after `dump` is completed.